File: AllToDo-iOS/AllToDo/Views/Components/RightSideControls.swift
import SwiftUI

struct RightSideControls: View {
    var compassRotation: Double
    var showHistoryMode: Bool
    var onHistoryClick: () -> Void
    var onNotificationClick: () -> Void
    var onLoginClick: () -> Void
    var onLocationClick: () -> Void
    var onZoomInClick: () -> Void
    var onZoomOutClick: () -> Void
    var onCompassClick: () -> Void
    var onExpandClick: () -> Void
    
    // [MODIFIED] Path Visualization Toggle
    var showActivePath: Bool
    var onRecordClick: () -> Void // Repurposed as toggle action
    
    var body: some View {
        Column(horizontalAlignment: .trailing) {
            // Top Group: Notification & Login
            HStack(spacing: 16) {
                ControlIcon(
                    iconName: showHistoryMode ? "calendar" : "clock.arrow.circlepath", 
                    onClick: onHistoryClick
                )
                
                ControlIcon(
                    iconName: "person.fill",
                    onClick: onLoginClick
                )
                .disabled(showHistoryMode)
                .opacity(showHistoryMode ? 0.3 : 1.0)
            }
            .padding(.bottom, 24)
            
            // Center Group: Location, Zoom, Compass, Path Toggle
            VStack(spacing: 16) {
                ControlIcon(
                    iconName: "location.fill",
                    onClick: onLocationClick
                )
                .disabled(showHistoryMode)
                .opacity(showHistoryMode ? 0.3 : 1.0)
                
                ControlIcon(
                    iconName: "plus",
                    onClick: onZoomInClick
                )
                .disabled(showHistoryMode)
                .opacity(showHistoryMode ? 0.3 : 1.0)
                
                ControlIcon(
                    iconName: "minus",
                    onClick: onZoomOutClick
                )
                .disabled(showHistoryMode)
                .opacity(showHistoryMode ? 0.3 : 1.0)
                
                // Compass
                let r = compassRotation.truncatingRemainder(dividingBy: 360)
                if abs(r) > 1.0 && abs(r) < 359.0 {
                    Button(action: onCompassClick) {
                        ZStack {
                            RoundedRectangle(cornerRadius: 12)
                                .fill(Color.allToDoGreen)
                                .frame(width: 48, height: 48)
                            
                            ZStack {
                                Path { path in
                                    path.move(to: CGPoint(x: 6, y: 0))
                                    path.addLine(to: CGPoint(x: 12, y: 18))
                                    path.addLine(to: CGPoint(x: 0, y: 18))
                                    path.closeSubpath()
                                }
                                .fill(Color.allToDoRed)
                                
                                Path { path in
                                    path.move(to: CGPoint(x: 0, y: 18))
                                    path.addLine(to: CGPoint(x: 12, y: 18))
                                    path.addLine(to: CGPoint(x: 6, y: 36))
                                    path.closeSubpath()
                                }
                                .fill(Color.white)
                                
                                Path { path in
                                    path.move(to: CGPoint(x: 6, y: 0))
                                    path.addLine(to: CGPoint(x: 12, y: 18))
                                    path.addLine(to: CGPoint(x: 6, y: 36))
                                    path.addLine(to: CGPoint(x: 0, y: 18))
                                    path.closeSubpath()
                                }
                                .stroke(Color.gray8, lineWidth: 1)
                            }
                            .frame(width: 12, height: 36)
                            .rotationEffect(.degrees(-compassRotation))
                        }
                    }
                    .buttonStyle(.plain)
                }
                
                // [MODIFIED] Path Visualization Toggle (Debugging Tool)
                Button(action: onRecordClick) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 12)
                            .fill(showActivePath ? Color.allToDoGreen : Color.white)
                            .opacity(0.8)
                            .frame(width: 48, height: 48)
                        
                        Image(systemName: "point.topleft.down.curvedto.point.bottomright.up")
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                            .frame(width: 24, height: 24)
                            .foregroundColor(.gray8)
                    }
                }
                .buttonStyle(.plain)
                .opacity(showActivePath ? 1.0 : 0.8)
            }
        }
    }
}

struct ControlIcon: View {
    var iconName: String
    var onClick: () -> Void
    var rotation: Double = 0
    var bgColor: Color = Color.allToDoGreen
    
    var body: some View {
        Button(action: onClick) {
            ZStack {
                RoundedRectangle(cornerRadius: 12)
                    .fill(bgColor.opacity(0.8))
                    .frame(width: 48, height: 48)
                
                Image(systemName: iconName)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
                    .frame(width: 24, height: 24)
                    .foregroundColor(.gray8)
                    .rotationEffect(.degrees(rotation))
            }
        }
        .buttonStyle(.plain)
    }
}

struct Column<Content: View>: View {
    var horizontalAlignment: HorizontalAlignment
    var content: () -> Content
    
    init(horizontalAlignment: HorizontalAlignment, @ViewBuilder content: @escaping () -> Content) {
        self.horizontalAlignment = horizontalAlignment
        self.content = content
    }
    
    var body: some View {
        VStack(alignment: horizontalAlignment, content: content)
    }
}

File: AllToDo-Android/app/src/main/java/kr/alltodo/ui/components/RightSideControls.kt
package kr.alltodo.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Explore
import androidx.compose.material.icons.filled.LocationOn
import androidx.compose.material.icons.filled.Login
import androidx.compose.material.icons.filled.MyLocation
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material.icons.filled.DirectionsWalk
import androidx.compose.material.icons.filled.Refresh // [FIX] Core Icon
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import kr.alltodo.ui.theme.*

import androidx.compose.material.icons.filled.ArrowDropUp
import kotlin.math.abs
import androidx.compose.material3.Text
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import androidx.compose.foundation.Image
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.Canvas
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke

// ... (Imports handled by context if needed, but adding specific ones)

@Composable
fun RightSideControls(
    modifier: Modifier = Modifier,
    compassRotation: Float = 0f,
    showHistoryMode: Boolean = false, // [NEW]
    onHistoryClick: () -> Unit = {}, // [NEW]
    onNotificationClick: () -> Unit = {},
    onLoginClick: () -> Unit = {},
    onLocationClick: () -> Unit = {},
    onZoomInClick: () -> Unit = {},
    onZoomOutClick: () -> Unit = {},
    onCompassClick: () -> Unit = {},
    isTracking: Boolean = false,
    showActivePath: Boolean = true,
    onToggleActivePath: () -> Unit = {},
    isDark: Boolean = androidx.compose.foundation.isSystemInDarkTheme()
) {
    val iconTint = if (isDark) Color.White else Color(0xFF333333)
    Column(
        modifier = modifier.padding(end = 16.dp),
        horizontalAlignment = Alignment.End
    ) {
        // [Item 1] My Info (Person)
        ControlIcon(
            icon = Icons.Default.Person,
            contentDescription = "My Info",
            onClick = onLoginClick,
            containerColor = AllToDoGreen.copy(alpha = 0.8f),
            iconTint = iconTint
        )

        Spacer(modifier = Modifier.height(24.dp)) // "--" Spacing (2 spaces)

        // [Item 2] Current Location Button
        ControlIcon(
            icon = Icons.Default.MyLocation,
            contentDescription = "Current Location",
            onClick = onLocationClick,
            containerColor = AllToDoGreen.copy(alpha = 0.8f),
            iconTint = iconTint
        )

        Spacer(modifier = Modifier.height(12.dp)) // "-" Spacing (1 space)

        // [Item 3] Attached Zoom Group
        Column(
            modifier = Modifier
                .width(48.dp)
                .clip(RoundedCornerShape(12.dp))
                .background(AllToDoGreen.copy(alpha = 0.8f))
        ) {
            Box(
                modifier = Modifier
                    .size(48.dp)
                    .clickable { onZoomInClick() },
                contentAlignment = Alignment.Center
            ) {
                Icon(Icons.Default.Add, contentDescription = "Zoom In", tint = iconTint, modifier = Modifier.size(24.dp))
            }
            // Thin divider
            Box(modifier = Modifier.fillMaxWidth().height(0.5.dp).background(iconTint.copy(alpha = 0.3f)))
            Box(
                modifier = Modifier
                    .size(48.dp)
                    .clickable { onZoomOutClick() },
                contentAlignment = Alignment.Center
            ) {
                Icon(Icons.Default.Remove, contentDescription = "Zoom Out", tint = iconTint, modifier = Modifier.size(24.dp))
            }
        }



        if (abs(compassRotation) > 0.01f) {
            Spacer(modifier = Modifier.height(12.dp)) // "-" Spacing (1 space)
            
            // [Item 4] Compass
            Box(
                modifier = Modifier
                    .size(48.dp)
                    .clip(RoundedCornerShape(12.dp))
                    .background(AllToDoGreen.copy(alpha = 0.8f))
                    .clickable { onCompassClick() },
                contentAlignment = Alignment.Center
            ) {
                Canvas(
                    modifier = Modifier
                        .width(10.dp)
                        .height(30.dp)
                        .rotate(-compassRotation)
                ) {
                    val widthMid = size.width / 2
                    val heightMid = size.height / 2

                    // Top (Red)
                    val topPath = Path().apply {
                        moveTo(widthMid, 0f)
                        lineTo(size.width, heightMid)
                        lineTo(0f, heightMid)
                        close()
                    }
                    drawPath(topPath, Color.Red) 

                    // Bottom (White)
                    val bottomPath = Path().apply {
                        moveTo(0f, heightMid)
                        lineTo(size.width, heightMid)
                        lineTo(widthMid, size.height)
                        close()
                    }
                    drawPath(bottomPath, Color.White)

                    // Outline
                    val outlinePath = Path().apply {
                        moveTo(widthMid, 0f)
                        lineTo(size.width, heightMid)
                        lineTo(widthMid, size.height)
                        lineTo(0f, heightMid)
                        close()
                    }
                    drawPath(outlinePath, iconTint, style = Stroke(width = 1f)) 
                }
            }
        }

        // [NEW] Active Path Toggle Button
        Spacer(modifier = Modifier.height(12.dp))
        Box(
            modifier = Modifier
                .size(48.dp)
                .clip(RoundedCornerShape(12.dp))
                .background(if (showActivePath) AllToDoGreen.copy(alpha = 0.8f) else Color.White.copy(alpha = 0.5f))
                .clickable { onToggleActivePath() },
            contentAlignment = Alignment.Center
        ) {
            Icon(
                painter = androidx.compose.ui.res.painterResource(id = kr.alltodo.R.drawable.ic_path_toggle),
                contentDescription = "Toggle Path",
                tint = if (showActivePath) iconTint else Color.Gray,
                modifier = Modifier.size(24.dp)
            )
        }


    }
}

@Composable
fun ControlIcon(
    icon: ImageVector,
    contentDescription: String,
    onClick: () -> Unit,
    iconModifier: Modifier = Modifier,
    enabled: Boolean = true,
    containerColor: Color = AllToDoGreen.copy(alpha = 0.7f),
    iconTint: Color = Gray8
) {
    Box(
        modifier = Modifier
            .size(48.dp)
            .clip(RoundedCornerShape(12.dp)) // Rounded Square
            .background(containerColor.copy(alpha = if (enabled) containerColor.alpha else 0.3f)) // Dim if disabled
            .clickable(enabled = enabled) { onClick() },
        contentAlignment = Alignment.Center
    ) {
        Icon(
            imageVector = icon,
            contentDescription = contentDescription,
            tint = iconTint.copy(alpha = if (enabled) 1f else 0.5f), // Dim icon
            modifier = iconModifier.size(24.dp)
        )
    }
}
